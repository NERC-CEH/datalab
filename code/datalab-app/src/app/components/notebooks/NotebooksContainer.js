import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { MODAL_TYPE_CONFIRMATION } from '../../constants/modaltypes';
import notebookActions from '../../actions/notebookActions';
import modalDialogActions from '../../actions/modalDialogActions';
import NotebookCards from './NotebookCards';
import PromisedContentWrapper from '../common/PromisedContentWrapper';
import notify from '../common/notify';

class NotebooksContainer extends Component {
  constructor(props, context) {
    super(props, context);
    this.openNotebook = this.openNotebook.bind(this);
    this.confirmDeleteNotebook = this.confirmDeleteNotebook.bind(this);
  }

  openNotebook(id) {
    return this.props.actions.getUrl(id)
      .then(payload => this.props.actions.openNotebook(payload.value.redirectUrl))
      .catch(err => notify.error('Unable to open Notebook'));
  }

  deleteNotebook = notebook =>
    Promise.resolve(this.props.actions.closeModalDialog())
      .then(() => this.props.actions.deleteNotebook(notebook))
      .then(this.props.actions.loadNotebooks)
      .then(() => notify.success('Notebook deleted'))
      .catch(err => notify.error('Unable to delete Notebook'));

  confirmDeleteNotebook(notebook) {
    this.props.actions.openModalDialog(MODAL_TYPE_CONFIRMATION, {
      title: `Delete ${notebook.displayName} Notebook`,
      body: `Would you like to delete the ${notebook.displayName} notebook? Any saved work will continue to be 
        stored in the shared drive.`,
      onSubmit: () => this.deleteNotebook(notebook),
      onCancel: this.props.actions.closeModalDialog,
    });
  }

  componentWillMount() {
    this.props.actions.loadNotebooks();
  }

  render() {
    return (
      <PromisedContentWrapper promise={this.props.notebooks}>
        <NotebookCards
          notebooks={this.props.notebooks.value}
          openNotebook={this.openNotebook}
          deleteNotebook={this.confirmDeleteNotebook}/>
      </PromisedContentWrapper>
    );
  }
}

NotebooksContainer.propTypes = {
  notebooks: PropTypes.shape({
    error: PropTypes.any,
    fetching: PropTypes.bool.isRequired,
    value: PropTypes.array.isRequired,
  }).isRequired,
  actions: PropTypes.shape({
    loadNotebooks: PropTypes.func.isRequired,
    getUrl: PropTypes.func.isRequired,
    openNotebook: PropTypes.func.isRequired,
    deleteNotebook: PropTypes.func.isRequired,
  }).isRequired,
};

function mapStateToProps({ notebooks }) {
  return { notebooks };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      ...notebookActions,
      ...modalDialogActions,
    }, dispatch),
  };
}

export { NotebooksContainer as PureNotebooksContainer }; // export for testing
export default connect(mapStateToProps, mapDispatchToProps)(NotebooksContainer);
