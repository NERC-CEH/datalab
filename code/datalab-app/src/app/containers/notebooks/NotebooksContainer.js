import React from 'react';
import StacksContainer from '../stacks/StacksContainer';
import { MODAL_TYPE_CREATE_NOTEBOOK } from '../../constants/modaltypes';

const CONTAINER_TYPE = 'analysis';
const TYPE_NAME = 'Notebook';

const NotebooksContainer = () => (
  <StacksContainer
    typeName={TYPE_NAME}
    containerType={CONTAINER_TYPE}
    dialogAction={MODAL_TYPE_CREATE_NOTEBOOK}
    formStateName='createNotebook' />);

export default NotebooksContainer;
