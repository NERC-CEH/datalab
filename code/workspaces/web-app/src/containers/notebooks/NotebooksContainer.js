import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { NOTEBOOK_CATEGORY } from 'common/src/config/images';
import StacksContainer from '../stacks/StacksContainer';
import { MODAL_TYPE_CREATE_NOTEBOOK, MODAL_TYPE_EDIT_NOTEBOOK } from '../../constants/modaltypes';
import EditNotebookForm from '../../components/notebooks/EditNotebookForm';
import { NOTEBOOK_TYPE_NAME, NOTEBOOK_TYPE_NAME_PLURAL } from './notebookTypeName';
import assetRepoActions from '../../actions/assetRepoActions';

export const CONTAINER_TYPE = NOTEBOOK_CATEGORY;
export const FORM_NAME = 'createNotebook';

const NotebooksContainer = ({ userPermissions }) => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(assetRepoActions.loadVisibleAssets());
  }, [dispatch]);

  return (
  <StacksContainer
    typeName={NOTEBOOK_TYPE_NAME}
    typeNamePlural={NOTEBOOK_TYPE_NAME_PLURAL}
    containerType={CONTAINER_TYPE}
    dialogAction={MODAL_TYPE_CREATE_NOTEBOOK}
    editDialogAction={MODAL_TYPE_EDIT_NOTEBOOK}
    formStateName={FORM_NAME}
    formComponent={EditNotebookForm}
    userPermissions={userPermissions}
  />
  );
};

NotebooksContainer.propTypes = {
  userPermissions: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default NotebooksContainer;
