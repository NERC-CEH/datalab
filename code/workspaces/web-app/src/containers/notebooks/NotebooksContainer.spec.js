import React from 'react';
import { useDispatch } from 'react-redux';
import { shallow } from 'enzyme';
import NotebooksContainer from './NotebooksContainer';

jest.mock('react-redux');
jest.mock('../stacks/StacksContainer', () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue(<>stacks container</>),
}));
jest.mock('../../components/notebooks/EditNotebookForm', () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue(<>notebook form</>),
}));

useDispatch.mockReturnValue(jest.fn().mockName('dispatch'));

describe('NotebooksContainer', () => {
  it('renders correctly passing correct props to children', () => {
    expect(shallow(<NotebooksContainer userPermissions={['expectedPermission']}/>)).toMatchSnapshot();
  });
});
