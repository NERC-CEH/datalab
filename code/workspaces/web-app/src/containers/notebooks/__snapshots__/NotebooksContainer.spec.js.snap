// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotebooksContainer renders correctly passing correct props to children 1`] = `
<mockConstructor
  containerType="ANALYSIS"
  dialogAction="MODAL_TYPE_CREATE_NOTEBOOK"
  editDialogAction="MODAL_TYPE_EDIT_NOTEBOOK"
  formComponent={[MockFunction]}
  formStateName="createNotebook"
  typeName="Notebook"
  typeNamePlural="Notebooks"
  userPermissions={
    Array [
      "expectedPermission",
    ]
  }
/>
`;
